rep-0149.rst

```diff
@@ -0,0 +1,1090 @@
+licenses.  Every license occurring in the source files should have
+a corresponding ``<license>`` tag.  For any explanatory text about
+licensing caveats, please use the ``<description>`` tag.
+
+Most common open-source licenses are described on the
+`OSI website <http://www.opensource.org/licenses/alphabetical>`_.
+
+Commonly used license strings:
+
+ - Apache 2.0
+ - BSD
+ - Boost Software License
+ - GPLv2
+ - GPLv3
+ - LGPLv2.1
+ - LGPLv3
+ - MIT
+ - Mozilla Public License Version 1.1
+
+<url> (multiple)
+----------------
+
+A Uniform Resource Locator for the package's website, bug tracker or
+source repository.
+
+It is a good idea to include ``<url>`` tags pointing users to these
+resources.  The website is commonly a wiki page on ``ros.org`` where
+users can find and update information about the package.
+
+Attributes
+''''''''''
+
+ ``type="TYPE"`` *(optional)*
+
+ The type should be one of the following identifiers: ``website``
+ (default), ``bugtracker`` or ``repository``.
+
+
+<author> (multiple)
+-------------------
+
+The name of a person who is an author of the package, as
+acknowledgement of their work and for questions.
+
+Attributes
+''''''''''
+
+ ``email="name@domain.tld"`` *(optional)*
+
+  Email address of author.
+
+
+Dependency tags
+---------------
+
+<build_depend> (multiple)
+'''''''''''''''''''''''''
+
+Declares a rosdep key or ROS package name that this package requires
+at build-time.  For system packages, the rosdep key will normally
+specify the "development" package, which frequently ends in ``"-dev"``.
+
+The ``build`` and ``buildtool`` dependencies are used to determine
+the build order of multiple packages.
+
+Attributes
+..........
+
+ All dependencies and relationships may restrict their applicability
+ to particular versions.  For each comparison operator an attribute
+ can be used.  Two of these attributes can be used together to
+ describe a version range.
+
+ ``version_lt="VERSION"`` *(optional)*
+
+ The dependency to the package is restricted to versions less than
+ the stated version number.
+
+ ``version_lte="VERSION"`` *(optional)*
+
+ The dependency to the package is restricted to versions less or
+ equal than the stated version number.
+
+ ``version_eq="VERSION"`` *(optional)*
+
+ The dependency to the package is restricted to a version equal than
+ the stated version number.
+
+ ``version_gte="VERSION"`` *(optional)*
+
+ The dependency to the package is restricted to versions greater or
+ equal than the stated version number.
+
+ ``version_gt="VERSION"`` *(optional)*
+
+ The dependency to the package is restricted to versions greater than
+ the stated version number.
+
+ ``condition="CONDITION_EXPRESSION"``
+
+ Every dependency can be conditional on a condition expression.
+ If the condition expression evaluate to "true" the dependency is being used
+ and considered as if it doesn't have a condition attribute.
+ If the condition expression evaluate to "false" the dependency is being
```

Comment:
Shouldn't there be a better definition of "evaluates"? In which language? In which semantics? In what context (e.g. what variables are available and how to set them)? Is it possible to compare against strings? Should strings be quoted? Why no "<", "<=", ">", ">=" ?

It would seem nice to me if there would be an agreement about this in-XML conditional expressions language and expressivity. Right now I know about this, about roslaunch, and about Xacro; maybe there are even more tools evaluating conditions inside XML files. It starts to be difficult to keep track about where I can use what...

Reply:
> Shouldn't there be a better definition of "evaluates"?

The string describes an expression which can be "evaluated" to a boolean value. I am not sure what kind of definition you would like to see. Can you please propose a text which clarifies your expectation.
 
> In which language?

Evaluating could happen in any language. E.g. [ament_package](https://github.com/ament/ament_package/pull/63/files#diff-27fa406e1d042c7052d279e1680b8635) provides a function in Python which parses the condition and can evaluate it based on some input variables.

> In which semantics? In what context (e.g. what variables are available and how to set them)?

Any tool evaluating these conditions has to define that. E.g. [ament_tools](https://github.com/ament/ament_tools/pull/168/files#diff-6d0e8f72766cf2fab285d79a89594f8fR196) passes all environment variables in. If a tool like rosdep needs to also support to be invoked without environment being set already then it could offer command line options like `--ros-version` (similar to the existing option `--ros-distro`.

> Is it possible to compare against strings? Should strings be quoted? Why no "<", "<=", ">", ">=" ?

All variables are interpreted as strings. The same as e.g. environment variables are being used in a shell. Since there is no typing beyond strings the meaning of  comparison operators is less useful. Since there wasn't a use case requiring this those operators where not specified in the current draft.

> It would seem nice to me if there would be an agreement about this in-XML conditional expressions language and expressivity. Right now I know about this, about roslaunch, and about Xacro; maybe there are even more tools evaluating conditions inside XML files. It starts to be difficult to keep track about where I can use what...

For complex conditions in XML would be one way. Imo the way roslaunch exposes those especially with the desire to express even more and more complicated conditions is not a sustainable way. The prototype of launch files in ROS 2 doesn't use XML for that reason. A script like Python is much better suited to express complex conditions and logic to decide if a node should be started or what the arguments / parameters should be. Pushing that complexity into XML is from my point of view not desired.

Reply:
Well, it seems to me that what's in this specification is just the syntax, but not the semantics of the expression.

There's e.g. not specified the meaning of the parentheses (of course it is "apparent", but unless written down, somebody may interpret it differently).

I suggest adding the following paragraph (after the list of tokens):

> An expression syntactically correct by the previous definition will be evaluated as follows. First, all variables are substituted by their values and treated as strings. All literals are also treated as strings. The resulting expression is then evaluated as a Python 3 interpreter would evaluate it.

*Beware, in Python `"1" != 1`! Empty string evaluates to False.*

Then I suggest adding a few hints about error handling. I just don't know which of these to add:

1. When a syntactically incorrect condition expression is encountered, the processing tool should stop with an error.
1. When a syntactically incorrect condition expression is encountered, the condition attribute is ignored and the processing tool should produce a warning or alert the developer (not necessarily the user).

Also, error handling for the variables should be defined:

1. When a variable is used in the condition expression, whose value is not known to the processing tool, it should stop processing and issue an error.
1. When a variable is used in the condition expression, whose value is not known to the processing tool, it should ignore the condition attribute and issue a warning to the developer.
1. When a variable is used in the condition expression, whose value is not known to the processing tool, it is substituted with and empty string (and a warning is issued to the developer).

As a final suggestion, an example could be given.

> As an example, a dependency might only be needed in ROS 1 build environment. Such dependency could read `<depend condition="$ROS_VERSION==1">roscpp</depend>`, where the value of `$ROS_VERSION` can be taken from the shell environment by catkin/ament.

I still think more comparison operators would be better, to be e.g. able to write `$ROS_DISTRO >= kinetic`. However I know that this would greatly increase the complexity of this specification. And it is not wise to just let there be any Python expression, because then even the C++ processing tools would need to emulate/access the Python interpreter. So, I am for a constrained syntax, but maybe not that much constrained.